HCStream.cpp

```diff
@@ -0,0 +1,269 @@
+    throw;
+  }
+
+}
+
+template <class T>
+void HCStream<T>::read_arrays(std::vector<T>& a, std::vector<T>& b, std::vector<T>& c)
+{
+  hc::copy(d_a,a.begin());
+  hc::copy(d_b,b.begin());
+  hc::copy(d_c,c.begin());
+}
+
+
+template <class T>
+void HCStream<T>::copy()
+{
+
+  hc::array_view<T,1> view_a = this->d_a;
+  hc::array_view<T,1> view_c = this->d_c;
+
+  try{
+    hc::completion_future future_kernel = hc::parallel_for_each(hc::extent<1>(array_size)
+                                , [=](hc::index<1> index) [[hc]] {
+                                  view_c[index] = view_a[index];
+								});
+    future_kernel.wait();
+  }
+  catch(std::exception& e){
+    std::cerr << __FILE__ << ":" << __LINE__ << "\t HCStream<T>::copy " << e.what() << std::endl;
+    throw;
+  }
+}
+
+template <class T>
+void HCStream<T>::mul()
+{
+
+  const T scalar = startScalar;
+  hc::array_view<T,1> view_b = this->d_b;
+  hc::array_view<T,1> view_c = this->d_c;
+
+  try{
+    hc::completion_future future_kernel = hc::parallel_for_each(hc::extent<1>(array_size)
+                                , [=](hc::index<1> i) [[hc]] {
+                                  view_b[i] = scalar*view_c[i];
+								});
+    future_kernel.wait();
+  }
+  catch(std::exception& e){
+    std::cerr << __FILE__ << ":" << __LINE__ << "\t HCStream<T>::copy " << e.what() << std::endl;
+    throw;
+  }
+}
+
+template <class T>
+void HCStream<T>::add()
+{
+
+
+  hc::array_view<T,1> view_a(this->d_a);
+  hc::array_view<T,1> view_b(this->d_b);
+  hc::array_view<T,1> view_c(this->d_c);
+
+  try{
+    hc::completion_future future_kernel = hc::parallel_for_each(hc::extent<1>(array_size)
+                                , [=](hc::index<1> i) [[hc]] {
+                                  view_c[i] = view_a[i]+view_b[i];
+								});
+    future_kernel.wait();
+  }
+  catch(std::exception& e){
+    std::cerr << __FILE__ << ":" << __LINE__ << "\t HCStream<T>::copy " << e.what() << std::endl;
+    throw;
+  }
+}
+
+template <class T>
+void HCStream<T>::triad()
+{
+
+  const T scalar = startScalar;
+  hc::array_view<T,1> view_a(this->d_a);
+  hc::array_view<T,1> view_b(this->d_b);
+  hc::array_view<T,1> view_c(this->d_c);
+
+  try{
+    hc::completion_future future_kernel = hc::parallel_for_each(hc::extent<1>(array_size)
+                                , [=](hc::index<1> i) [[hc]] {
+                                  view_a[i] = view_b[i] + scalar*view_c[i];
+								});
+    future_kernel.wait();
+  }
+  catch(std::exception& e){
+    std::cerr << __FILE__ << ":" << __LINE__ << "\t HCStream<T>::copy " << e.what() << std::endl;
+    throw;
+  }
+}
+
+template <class T>
+T HCStream<T>::dot()
+{
+   //implementation adapted from
+    //https://ampbook.codeplex.com/SourceControl/latest
+    // ->Samples/CaseStudies/Reduction
+    // ->CascadingReduction.h
+
+    static constexpr std::size_t n_tiles = 64;
```

Comment:
Where does this number come from? Is this something that could/should be determined dynamically from the device specification?